### ISSUE TITLE
TypeNameHandling.All in JSON settings of the storage providers

### ISSUE THREAD

**Issue by DixonDs** [dotnet/orleans#3045]
Will it work if we don't put "$type" into JSON-serialized grain states when built-in Storage providers are used such as Azure Table and Blob storage providers? Since we know the current grain state type, I imagine, that should be enough to deserialize. That would allow to rename grain state types without breaking compatibility with old serialized states.

Somehow, changing grain state type names didn't break serialization in 1.3.1. We had some old grain states with renamed state types in codebase months ago, and it was working until we migrated to 1.4.1.

**Comment by ReubenBond** [comment]
I haven't tested, @DixonDs, but I believe it will work, since we do use that setting in our storage providers.

**Comment by DixonDs** [comment]
@ReubenBond We use `TypeNameHandling.All`, while I propose something like `TypeNameHandling.None`. I will check myself later if that works, I was just hoping for a quick answer if that's possible or not :)

**Comment by ReubenBond** [comment]
Using `TypeNameHandling.None` is fine if all types are statically known, i.e, JSON.NET never encounters something like `MyBaseClass` where you have some specific subclass at runtime, or `object`, where the concrete class just isn't known at all.

**Comment by ReubenBond** [comment]
Does this answer your question, @DixonDs ?

**Comment by DixonDs** [comment]
@ReubenBond The goal I was trying to achieve here was to have more forgiving serialization settings so that refactorings like grain state type renames would not break serialization. Based on your response, `TypeNameHandling.None` can't be applied in the general case, so in some sense the original question was answered. I guess it would be beneficial to have some means to change JSON settings through the configuration and/or make it possible to change serializer easily in scope of the specific storage provider (related to https://github.com/dotnet/orleans/issues/895)

**Comment by ReubenBond** [comment]
That will be more easily achieved soon (as Sergey mentions in that thread) :) If you can modify the JSON.NET config (which we don't allow yet), then you can provide a [`SerializationBinder`](http://www.newtonsoft.com/json/help/html/SerializationSettings.htm#Binder) and map between old & new type names in there.

If that sounds like it would solve this for you, would you mind opening another issue (link back to here) and after #895 is resolved, we can look into it?

**Comment by DixonDs** [comment]
@ReubenBond sure, we can close this one
